### ISSUE TITLE
[Proposal]Alias dispatch keys

### ISSUE THREAD

**Issue by ailzhang** [pytorch/pytorch#44680]
Only attaching the proposal here, will add the colab from @bhosmer once I finish updating it. 

## Problem:

* Consider _convolution implementation (runs before autograd, selects cpu/cuda implementation to do the actual work). Today, this is registered to both the CPU and Autograd keys via CatchAll mechanism, which means that we will use this implementation even if autograd is disabled (e.g., by NonVariableTypeMode). When we get rid of CatchAll, we still need to make sure these implementations get registered to CPU key. So for example, _convolution would have to be registered to CPU, CUDA, AutogradCPU, AutogradCUDA. This is pretty wordy and a bit difficult for people to understand.

Solution: Offer synthetic dispatch keys which map to multiple dispatch keys, so that you can conveniently refer to “CPU, CUDA, AutogradCPU, AutogradCUDA” all in one go. We call these *alias dispatch keys.*

## Proposal:

* Distinguish between runtime dispatch keys (keys that actually show up in the dispatch table) and alias dispatch keys (dispatch keys which map to other dispatch keys). Alias keys have precedence, but they are always lower precedence than runtime keys. *New keys bolded.*
    * Runtime keys: CPU, CUDA, XLA, *AutogradCPU*, *AutogradCUDA*, AutogradXLA (previously PreAutogradXLA)
    * Alias keys: *AutogradDispatchCPUOrCUDA* > *Math* > *Autograd* (previously a runtime key, now alias)
* When you register a kernel for an alias key, you register the kernel for some set of dispatch keys.
    * AutogradDispatchCPUOrCuda => CPU, CUDA, AutogradCPU, AutogradCUDA
    * Autograd => AutogradCPU, AutogradCUDA, AutogradXLA
    * Math => CPU, CUDA, XLA, AutogradCPU, AutogradCUDA, AutogradXLA (e.g., DefaultBackend and Autograd)
* If a runtime dispatch key has multiple kernels, which kernel wins is done based on precedence of alias keys (but runtime keys always have precedence).
    * Example: convolution. Kernels are registered to AutogradDispatchCPUOrCuda (at::native::_convolution), Autograd (derivatives.yaml), XLA (XLA implementation)
        * AutogradXLA => derivatives.yaml (only key that maps to this key)
        * AutogradCPU => at::native::_convolution (beats derivatives.yaml because AutogradDispatchCPUOrCuda > Autograd)
* Then there’s some mumbling you have to do about Math versus backend key: when there’s a direct registration to backend key X, AutogradX key should skip picking up Math kernel.

### What are the use cases? 

Consider operator at::sign which currently 

* has no dispatch section in native_functions.yaml => Use TypeDefault and registered as CatchAll
    * note this kernel works for all backends as it just redirect calls to at::sign_out which has a dispatch section in native_functions.yaml. 
    * at::sign_out actually doesn’t have derivative defined.
* It has an entry in derivatives.yaml.

The problem happens when we try to move at::sign from catchAll to Math key, since Math > Autograd, Math kernel would just override the Autograd kernel which was expected to be used. (Autograd > catchAll) 

### Proposed solution?

This requires us to have a way in native_functions.yaml to say 

* hey please only use TypeDefault::sign for inference, and use kernel from Autograd for training if available

solution: DefaultBackend > Math > Autograd
 for every op that has an entry in derivatives.yaml but no explicit dispatch section in native_functions.yaml, register TypeDefault::op to DefaultBackend. VariableType::op to Autograd. 

### What does it mean to current native_functions.yaml? 

Ops like at::sign is currently registered to catchAll, they should be registered as DefaultBackend (by adding DefaultBackend in native_functions.yaml) before we can move catchAll kernels to Math. 

Ops we need to move has the following properties:

1. it doesn’t have dispatch section in native_functions.yaml
2. its kernel works for all backends (calling ops using at::) .
3. its implementation uses an op doesn’t have autograd. (like _out function) 
4. it either has an entry in derivatives.yaml (or it’s inplace version of the op), or has a manual Autograd kernel. 

In general this improves native_functions.yaml that:
by look at native_functions.yaml,

* we know exactly which backend the kernel support (it was not accurate due to dispatchStub and is_sparse() etc), 

* where does the training support for this op should come from for each backend.

**Comment by zou3519** [comment]
Some noob questions: I'm still a bit confused as to what the Math dispatch key does. Let's take a concrete example: addr.
It has the following native_functions.yaml entry:
```
- func: addr(Tensor self, Tensor vec1, Tensor vec2, *, Scalar beta=1, Scalar alpha=1) -> Tensor
  use_c10_dispatcher: full
  variants: function, method
  dispatch:
    CPU, CUDA: addr
    Math: math_addr
```
and the following derivatives.yaml entry:
```
- name: addr(Tensor self, Tensor vec1, Tensor vec2, *, Scalar beta=1, Scalar alpha=1) -> Tensor
  self: maybe_multiply(grad, beta)
  vec1: grad.mv(vec2) * alpha
  vec2: grad.t().mv(vec1) * alpha
```
Now, a few questions:
1. On CPU and CUDA, addr has an implementation for the AutogradCPU / AutogradCUDA dispatch keys. Do those get automatically generated from derivatives.yaml?
2. How does a Tensor with a non-standard "backend" actually hit the Math dispatch key? E.g. if you have a Tensor with XLA on it, what dispatch keys does it have and how does that hit the Math kernel? I think the difficulty in visualizing this is that `Math` isn't actually a dispatch key, it maps to a set of dispatch keys and there is some priority defined.
3. There have been ideas floating around to implement a "reference implementation" for things like nn.functional functions. I *think* one way to approach that is to write a `Math` implementation for each nn.functional (like relu, elu, celu, etc). Would that help things like torch/xla have better coverage, or does something like torch/xla do something special for things like relu/elu/celu?

Thanks in advance.

**Comment by ailzhang** [comment]
@zou3519 Great questions! 
1. Yes they got generated from derivatives.yaml and got populated to AutogradCPU/AutogradCUDA entries in dispatch table. 
2. Yea Math maps to a set of backend keys and it has lower priority than the explicit registration to those keys. For example, in the addr case above, `math_addr` will be populated to both `AutogradXLA` and `XLA` if XLA doesn't have a direct registration for `addr`. (They have the option to register but they can use math_addr if they want). And since we do have direct registrations for `CPU/CUDA/AutogradCPU/AutogradCUDA` `math_addr` won't be populated into those entries. 
3. Yea `Math` is designed to act as a "default implementation" for all backends, the main goal is to minimize the set of operators backends "have to" implement. Each backend can do special registrations based on their hardware but they'll have the option to use Math kernels as well. The main missing piece in the framework for Math is the testing for the backward pass, we currently only compare results of math/cpu kernel for the forward pass. 
Let me know if you have any followup questions!

**Comment by ezyang** [comment]
This is all implemented!
